core/src/main/scala/io/finch/Endpoint.scala

```diff
@@ -238,127 +239,136 @@ trait Endpoint[A] { self =>
   }
 
   /**
-   * Composes this endpoint with another in such a way that coproducts are flattened.
-   */
-  final def :+:[B](that: Endpoint[B])(implicit a: Adjoin[B :+: A :+: CNil]): Endpoint[a.Out] = {
+    * Composes this endpoint with another in such a way that coproducts are flattened.
+    */
+  final def :+:[B](that: Endpoint[F, B])(implicit a: Adjoin[B :+: A :+: CNil]): Endpoint[F, a.Out] = {
     val left = that.map(x => a(Inl[B, A :+: CNil](x)))
     val right = self.map(x => a(Inr[B, A :+: CNil](Inl[A, CNil](x))))
 
     left.coproduct(right)
   }
 
   /**
-   * Converts this endpoint to a Finagle service `Request => Future[Response]` that serves JSON.
-   *
-   * Consider using [[Bootstrap]] instead.
-   */
+    * Converts this endpoint to a Finagle service `Request => Future[Response]` that serves JSON.
+    *
+    * Consider using [[io.finch.Bootstrap]] instead.
+    */
   final def toService(implicit
-    tr: ToResponse.Aux[A, Application.Json],
-    tre: ToResponse.Aux[Exception, Application.Json]
-  ): Service[Request, Response] = toServiceAs[Application.Json]
+                      tr: ToResponse.Aux[A, Application.Json],
+                      tre: ToResponse.Aux[Exception, Application.Json]
+                     ): Service[Request, Response] = toServiceAs[Application.Json]
 
   /**
-   * Converts this endpoint to a Finagle service `Request => Future[Response]` that serves custom
-   * content-type `CT`.
-   *
-   * Consider using [[Bootstrap]] instead.
-   */
+    * Converts this endpoint to a Finagle service `Request => Future[Response]` that serves custom
+    * content-type `CT`.
+    *
+    * Consider using [[io.finch.Bootstrap]] instead.
+    */
   final def toServiceAs[CT <: String](implicit
-    tr: ToResponse.Aux[A, CT],
-    tre: ToResponse.Aux[Exception, CT]
-  ): Service[Request, Response] = Bootstrap.serve[CT](this).toService
+                                      tr: ToResponse.Aux[A, CT],
+                                      tre: ToResponse.Aux[Exception, CT]
+                                     ): Service[Request, Response] = Bootstrap.serve[CT](this).toService
 
   /**
-   * Recovers from any exception occurred in this endpoint by creating a new endpoint that will
-   * handle any matching throwable from the underlying future.
-   */
-  final def rescue[B >: A](pf: PartialFunction[Throwable, Future[Output[B]]]): Endpoint[B] =
-    transform(foa => foa.rescue(pf))
+    * Recovers from any exception occurred in this endpoint by creating a new endpoint that will
+    * handle any matching throwable from the underlying future.
+    */
+  final def rescue(pf: PartialFunction[Throwable, F[Output[A]]]): Endpoint[F, A] =
+    transform(foa => foa.handleErrorWith(pf))
```

Comment:
Due to the fact that `F[_]` is invariant and `MonadError.handleErrorWith` doesn't give any lower bounds, we can't use `Output[B]` where `B >: A` anymore. That's quite sad